src/target/riscv/riscv.c

```diff
@@ -604,90 +606,349 @@ static int maybe_add_trigger_t1(struct target *target, struct trigger *trigger)
-	if (ret != ERROR_OK)
-		return ret;
-	tdata1 = set_field(tdata1, CSR_MCONTROL_MATCH, CSR_MCONTROL_MATCH_NAPOT);
-	tdata1 = set_field(tdata1, CSR_MCONTROL_CHAIN, CSR_MCONTROL_CHAIN_DISABLED);
-	ret = set_trigger(target, idx, tdata1, tdata2, CSR_MCONTROL_MASKMAX(riscv_xlen(target)));
-	if (ret != ERROR_OK) {
-		if (ret == ERROR_TARGET_RESOURCE_NOT_AVAILABLE)
-			goto MATCH_GE_LT; /* Fallback to chained MATCH using GT and LE */
-		return ret;
-	}
-	r->trigger_unique_id[idx] = trigger->unique_id;
-	return ERROR_OK;
+static riscv_reg_t set_size_s8bit_t2(void)
+{
+	riscv_reg_t tdata1 = 0;
+	tdata1 = set_field(tdata1, CSR_MCONTROL_SIZELO, CSR_MCONTROL_SIZELO_8BIT & 3);
+	tdata1 = set_field(tdata1, CSR_MCONTROL_SIZEHI, (CSR_MCONTROL_SIZELO_8BIT >> 2) & 3);
+	return tdata1;
+}
 
-MATCH_GE_LT:
-	ret = find_trigger(target, CSR_TDATA1_TYPE_MCONTROL, true, &idx);
-	if (ret != ERROR_OK)
-		return ret;
-	tdata1 = set_field(tdata1, CSR_MCONTROL_MATCH, CSR_MCONTROL_MATCH_GE);
+static riscv_reg_t set_size_s8bit_t6(void)
+{
+	riscv_reg_t tdata1 = 0;
+	tdata1 = set_field(tdata1, CSR_MCONTROL6_SIZE, CSR_MCONTROL6_SIZE_8BIT);
+	return tdata1;
+}
+
+static riscv_reg_t set_chain_enable_t2(void)
+{
+	riscv_reg_t tdata1 = 0;
 	tdata1 = set_field(tdata1, CSR_MCONTROL_CHAIN, CSR_MCONTROL_CHAIN_ENABLED);
-	tdata2 = trigger->address;
-	ret = set_trigger(target, idx, tdata1, tdata2, 0);
-	if (ret != ERROR_OK) {
-		if (ret == ERROR_TARGET_RESOURCE_NOT_AVAILABLE)
-			goto MATCH_EQUAL; /* Fallback to EQUAL MATCH */
-		return ret;
-	}
-	tdata1 = set_field(tdata1, CSR_MCONTROL_MATCH, CSR_MCONTROL_MATCH_LT);
+	return tdata1;
+}
+
+static riscv_reg_t set_chain_enable_t6(void)
+{
+	riscv_reg_t tdata1 = 0;
+	tdata1 = set_field(tdata1, CSR_MCONTROL6_CHAIN, CSR_MCONTROL6_CHAIN_ENABLED);
+	return tdata1;
+}
+
+static riscv_reg_t set_chain_disable_t2(void)
+{
+	riscv_reg_t tdata1 = 0;
 	tdata1 = set_field(tdata1, CSR_MCONTROL_CHAIN, CSR_MCONTROL_CHAIN_DISABLED);
-	tdata2 = trigger->address + trigger->length;
-	ret = set_trigger(target, idx + 1, tdata1, tdata2, 0);
-	if (ret != ERROR_OK) {
-		/* Undo the setting of the previous trigger*/
-		set_trigger(target, idx, 0, 0, CSR_MCONTROL_MASKMAX(riscv_xlen(target)));
-		if (ret == ERROR_TARGET_RESOURCE_NOT_AVAILABLE)
-			goto MATCH_EQUAL; /* Fallback to EQUAL MATCH */
-		return ret;
-	}
-	r->trigger_unique_id[idx] = trigger->unique_id;
-	r->trigger_unique_id[idx + 1] = trigger->unique_id;
-	return ERROR_OK;
+	return tdata1;
+}
 
-MATCH_EQUAL:
-	ret = find_trigger(target, CSR_TDATA1_TYPE_MCONTROL, false, &idx);
-	if (ret != ERROR_OK)
-		return ret;
+static riscv_reg_t set_chain_disable_t6(void)
+{
+	riscv_reg_t tdata1 = 0;
+	tdata1 = set_field(tdata1, CSR_MCONTROL6_CHAIN, CSR_MCONTROL6_CHAIN_DISABLED);
+	return tdata1;
+}
+
+static riscv_reg_t set_match_napot_t2(void)
+{
+	riscv_reg_t tdata1 = 0;
+	tdata1 = set_field(tdata1, CSR_MCONTROL_MATCH, CSR_MCONTROL_MATCH_NAPOT);
+	return tdata1;
+}
```

Comment:
All these simple 3-line functions can be inlined directly into the proposed `fill_trigger_request_info_*` functions. It would simply get reduced to a single line, for example: `tdata1 = set_field(tdata1, CSR_MCONTROL_MATCH, CSR_MCONTROL_MATCH_NAPOT);`

Reply:
Agreed. I appreciate that you're trying to avoid duplicating code and using symbolic names, but it feels unnecessarily complex to make so many tiny functions.

Reply:
Addressed.